src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIP.java

```diff
@@ -0,0 +1,242 @@
+     
+    /**
+     * {@inheritDoc}
+     * This PIP adds a {@value #ATTR_CA_POLICY_NAMES} attribute to the
+     * corresponding subjects. The value(s) of this attribute are the short
+     * names of all the {@value PolicyNamesPIPCache#FILE_SFX} files that match
+     * the value of the {@value ATTR_X509_ISSUER} attribute.
+     * @param request the incoming request.
+     * @throws PIPProcessingException in case of errors.
+     * @return boolean: true when attribute has been populated, false otherwise.
+     */
+    public boolean populateRequest(Request request)
+	throws PIPProcessingException
+    {
+	long t0=System.nanoTime();
+	boolean pipprocessed=false;
+	String issuerdn=null;
+
+	// Get all subjects from the request, should be at least one, warn
+	// when there are more than 1
+	Set<Subject> subjects = request.getSubjects();
+	if (subjects.isEmpty())	{
+	    log.error("Request has no subjects");
+	    throw new PIPProcessingException("No subject found in request");
+	}
+	if (subjects.size()>1)
+	    log.warn("Request has "+subjects.size()+
+		     " subjects, taking first match");
+
+	// Loop over all subjects
+	for (Subject subject : subjects) {
+	    // Loop over all attributes, looking for ATTR_X509_ISSUER
+	    Set<Attribute> attributes = subject.getAttributes();
+	    for (Attribute attr: attributes) {
+		if (ATTR_X509_ISSUER.equals(attr.getId())) {
+		    // Take first value (it should be singlevalued)
+		    Object tmp = attr.getValues().iterator().next();
+		    issuerdn = (tmp!=null ? tmp.toString() : null);
+		    break;
+		}
+	    }
+
+	    // Did we find the issuer attribute?
+	    if (issuerdn==null)	{
+		log.info("Subject has no or invalid "+ATTR_X509_ISSUER+
+			 " attribute set");
+		continue;
+	    }
+
+	    // Look for the issuerdn in the .info files
+	    String[] policynames=new String[0];
+	    try {
+		policynames=policyNamesCache.findIssuerDN(issuerdn);
+	    } catch (IOException e)	{
+		log.error("I/O error reading info files: "+e.getMessage());
+		throw new PIPProcessingException(
+		    "I/O error reading info files: "+e.getMessage());
+	    }
+
+	    // Log total number of matching policies
+	    log.debug("Found "+policynames.length+" matching policies");
+
+	    // Check that we found any names
+	    if (policynames.length==0)	{
+		log.info("No matching info file for this subject");
+		continue;
+	    }
+
+	    // Create new attribute and add the policy names
+	    Attribute attr_policynames =
+		new Attribute(attribute_name,
+			      Attribute.DT_STRING);
+	    Set<Object> values = attr_policynames.getValues();
+	    for (int i=0; i<policynames.length; i++)
+		values.add(policynames[i]);
+
+	    // Add to the current subject
+	    attributes.add(attr_policynames);
+	    pipprocessed=true;
+	    log.debug("Added attribute \""+attribute_name+"\"");
+	}
+
+	// Log statistics
+	log.debug("PIP parsing took "+(System.nanoTime()-t0)/1000000.0+" msec");
```

Comment:
That would be reasonable, but rounds down (truncates) to a whole msec. Hence I prefer to do the division.